send_review_notification/github_services.py

```diff
@@ -0,0 +1,266 @@
+    assigned to reviewers.
+    """
+
+    pr_url = PULL_REQUESTS_URL_TEMPLATE.format(org_name, repository)
+    reviewer_to_assigned_prs: (
+        DefaultDict[str, List[github_domain.PullRequest]]) = (
+        collections.defaultdict(list))
+
+    page_number = 1
+    while True:
+        logging.info('Fetching Pull requests')
+        params: Dict[str, Union[str, int]] = {
+            'page': page_number, 'per_page': 100, 'status': 'open'
+        }
+        response = requests.get(
+            pr_url,
+            params=params,
+            headers=_get_request_headers()
+        )
+        response.raise_for_status()
+        pr_subset = response.json()
+
+        if len(pr_subset) == 0:
+            break
+        page_number += 1
+
+        pull_requests = [
+            github_domain.PullRequest.from_github_response(pull_request)
+            for pull_request in pr_subset]
+        update_assignee_timestamp(org_name, repository, pull_requests)
+        for pull_request in pull_requests:
+            if not pull_request.is_reviewer_assigned():
+                continue
+            for reviewer in pull_request.assignees:
+                pending_review_time = (
+                    datetime.datetime.now(datetime.timezone.utc) -
+                    reviewer.timestamp)
+                if (reviewer.name != pull_request.author) and (
+                        pending_review_time >=
+                        datetime.timedelta(hours=wait_hours)
+                    ):
+                    reviewer_to_assigned_prs[reviewer.name].append(pull_request)
+    return reviewer_to_assigned_prs
+
+
+# Here we use type Any because the response we get from the api call is hard
+# to annotate in a typedDict, Hence used type Any.
+def __process_activity(
+    pull_request: github_domain.PullRequest,
+    event: Dict[str, Any]
+) -> None:
+    """Process activity and updates assignee timestamps."""
+    if event['event'] != 'assigned':
+        return
+
+    assignee = pull_request.get_assignee(event['assignee']['login'])
+    event_timestamp = parser.parse(event['created_at'])
+    if assignee:
+        assignee.set_timestamp(max([assignee.timestamp, event_timestamp]))
+
+
+def update_assignee_timestamp(
+    org_name: str,
+    repository: str,
+    pr_list: List[github_domain.PullRequest]
+) -> None:
+    """Fetches PR timeline and updates assignment timestamp."""
+    for pull_request in pr_list:
+        pr_number = pull_request.number
+        activity_url = ISSUE_TIMELINE_URL_TEMPLATE.format(
+            org_name, repository, pr_number)
+
+        page_number = 1
+        while True:
+            logging.info('Fetching PR #%s timeline', pr_number)
+            response = requests.get(
+                activity_url,
+                params={'page': page_number, 'per_page': 100},
+                headers={
+                    'Accept': 'application/vnd.github.mockingbird-preview+json',
```

Comment:
This looks like the accept header for a preview API feature, but it looks like timelines are no longer in preview mode: https://docs.github.com/en/rest/issues/timeline?apiVersion=2022-11-28#list-timeline-events-for-an-issue

Reply:
updated